### PR TITLE
Fix issue #4176

### PR DESCRIPTION
**Pull request by AllenX2018**
Fix issue #4176 

1. Refactor `Router` with option pattern
2. Add `SessionOn` option for `web.Router`

### CODE REVIEW

**Review by flycash — COMMENTED** [review]
Good! I will check it and merge it after release 2.0.0

File: server/web/router.go [hunk 1]
```diff
@@ -171,40 +186,67 @@ func NewControllerRegisterWithCfg(cfg *Config) *ControllerRegister {
...
+
+func (p *ControllerRegister) addWithMethodParams(pattern string, c ControllerInterface, methodParams []*param.MethodParam, opts ...ControllerOptions) {
+	reflectVal := reflect.ValueOf(c)
+	t := reflect.Indirect(reflectVal).Type()
+
 	route := &ControllerInfo{}
 	route.pattern = pattern
-	route.methods = methods
 	route.routerType = routerTypeBeego
+	route.sessionOn = BConfig.WebConfig.Session.SessionOn
```
Comment (flycash) [inline server/web/router.go:249]: ControllerRegister has a field name 'cfg', please using it.

File: server/web/router.go [hunk 2]
```diff
@@ -742,7 +778,16 @@ func (p *ControllerRegister) serveHttp(ctx *beecontext.Context) {
 	}
 
 	// session init
-	if p.cfg.WebConfig.Session.SessionOn {
+	currentSessionOn = BConfig.WebConfig.Session.SessionOn
+	originRouterInfo, originFindRouter = p.FindRouter(ctx)
+	if originFindRouter {
+		if !currentSessionOn && originRouterInfo.sessionOn {
```
Comment (flycash) [inline server/web/router.go:784]: could we calcaulte the value of `sessionOn` when we create this route?

**Review by flycash — COMMENTED** [review]
Do not use global config BConfig anymore, it should only be used in initialization.

Please use ControllerRegister.cfg instead.

File: server/web/router.go (hunk 1 above)
Comment (AllenX2018) [reply server/web/router.go:249]: OK

File: server/web/router.go (hunk 2 above)
Comment (AllenX2018) [reply server/web/router.go:784]: Good idea. I'll try it.

**Review by jianzhiyao — APPROVED** [review]
checked

**Pull request merged**